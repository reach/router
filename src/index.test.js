/* eslint-disable react/prop-types */
import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { renderToString } from "react-dom/server";

import {
  createHistory,
  createMemorySource,
  Router,
  LocationProvider,
  Link,
  Match,
  Redirect,
  isRedirect,
  ServerLocation
} from "./index";

let snapshot = ({ pathname, element }) => {
  let testHistory = createHistory(createMemorySource(pathname));
  let wrapper = renderer.create(
    <LocationProvider history={testHistory}>{element}</LocationProvider>
  );
  const tree = wrapper.toJSON();
  expect(tree).toMatchSnapshot();
  return tree;
};

let runWithNavigation = (element, pathname = "/") => {
  let history = createHistory(createMemorySource(pathname));
  let wrapper = renderer.create(
    <LocationProvider history={history}>{element}</LocationProvider>
  );
  const snapshot = string => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  };
  return { history, snapshot, wrapper };
};

let Home = () => <div>Home</div>;
let Dash = ({ children }) => <div>Dash {children}</div>;
let Group = ({ groupId, children }) => (
  <div>
    Group: {groupId}
    {children}
  </div>
);
let PropsPrinter = props => <pre>{JSON.stringify(props, null, 2)}</pre>;
let Reports = ({ children }) => <div>Reports {children}</div>;
let AnnualReport = () => <div>Annual Report</div>;

describe("smoke tests", () => {
  it(`renders the root component at "/"`, () => {
    snapshot({
      pathname: "/",
      element: (
        <Router>
          <Home path="/" />
          <Dash path="/dash" />
        </Router>
      )
    });
  });

  it("renders at a path", () => {
    snapshot({
      pathname: "/dash",
      element: (
        <Router>
          <Home path="/" />
          <Dash path="/dash" />
        </Router>
      )
    });
  });
});

describe("passed props", () => {
  it("parses dynamic segments and passes to components", () => {
    snapshot({
      pathname: "/group/123",
      element: (
        <Router>
          <Home path="/" />
          <Group path="/group/:groupId" />
        </Router>
      )
    });
  });

  it("passes the matched URI to the component", () => {
    snapshot({
      pathname: "/groups/123/users/456",
      element: (
        <Router>
          <PropsPrinter path="/groups/:groupId/users/:userId" />
        </Router>
      )
    });
  });

  it("shadows params in nested paths", () => {
    snapshot({
      pathname: `/groups/burger/groups/milkshake`,
      element: (
        <Router>
          <Group path="groups/:groupId">
            <Group path="groups/:groupId" />
          </Group>
        </Router>
      )
    });
  });

  it("parses multiple params when nested", () => {
    const Group = ({ groupId, children }) => (
      <div>
        {groupId}
        {children}
      </div>
    );
    const User = ({ userId, groupId }) => (
      <div>
        {groupId} - {userId}
      </div>
    );
    snapshot({
      pathname: `/group/123/user/456`,
      element: (
        <Router>
          <Group path="group/:groupId">
            <User path="user/:userId" />
          </Group>
        </Router>
      )
    });
  });
});

describe("route ranking", () => {
  const Root = () => <div>Root</div>;
  const Groups = () => <div>Groups</div>;
  const Group = ({ groupId }) => <div>Group Id: {groupId}</div>;
  const MyGroup = () => <div>MyGroup</div>;
  const MyGroupsUsers = () => <div>MyGroupUsers</div>;
  const Users = () => <div>Users</div>;
  const UsersSplat = ({ splat }) => <div>Users Splat: {splat}</div>;
  const User = ({ userId, groupId }) => (
    <div>
      User id: {userId}, Group Id: {groupId}
    </div>
  );
  const Me = () => <div>Me!</div>;
  const MyGroupsAndMe = () => <div>Mine and Me!</div>;
  const Fiver = ({ one, two, three, four, five }) => (
    <div>
      Fiver {one} {two} {three} {four} {five}
    </div>
  );

  const element = (
    <Router>
      <Root path="/" />
      <Groups path="/groups" />
      <Group path="/groups/:groupId" />
      <MyGroup path="/groups/mine" />
      <Users path="/groups/:groupId/users" />
      <MyGroupsUsers path="/groups/mine/users" />
      <UsersSplat path="/groups/:groupId/users/*" />
      <User path="/groups/:groupId/users/:userId" />
      <Me path="/groups/:groupId/users/me" />
      <MyGroupsAndMe path="/groups/mine/users/me" />
      <Fiver path="/:one/:two/:three/:four/:five" />
    </Router>
  );

  test("/", () => {
    snapshot({ element, pathname: "/" }); // Root
  });
  test("/groups", () => {
    snapshot({ element, pathname: "/groups" }); // Groups
  });
  test("/groups/123", () => {
    snapshot({ element, pathname: "/groups/123" }); // Group
  });
  test("/groups/mine", () => {
    snapshot({ element, pathname: "/groups/mine" }); // MyGroup
  });

  test("/groups/123/users", () => {
    snapshot({ element, pathname: "/groups/123/users" }); // Users
  });

  test("/groups/mine/users", () => {
    snapshot({ element, pathname: "/groups/mine/users" }); // MyGroupsUsers
  });

  test("/groups/123/users/456", () => {
    snapshot({ element, pathname: "/groups/123/users/456" }); // User
  });

  test("/groups/123/users/me", () => {
    snapshot({ element, pathname: "/groups/123/users/me" }); // Me
  });

  test("/groups/123/users/a/bunch/of/junk", () => {
    snapshot({
      element,
      pathname: "/groups/123/users/a/bunch/of/junk"
    }); // UsersSplat
  });

  test("/groups/mine/users/me", () => {
    snapshot({ element, pathname: "/groups/mine/users/me" }); // MyGroupsAndMe
  });

  test("/one/two/three/four/five", () => {
    snapshot({ element, pathname: "/one/two/three/four/five" }); // Fiver
  });
});

describe("nested rendering", () => {
  it("renders a nested path", () => {
    snapshot({
      pathname: "/dash/reports",
      element: (
        <Router>
          <Home path="/" />
          <Dash path="/dash">
            <Reports path="reports" />
          </Dash>
        </Router>
      )
    });
  });

  it("renders a really nested path", () => {
    snapshot({
      pathname: "/dash/reports/annual",
      element: (
        <Router>
          <Home path="/" />
          <Dash path="/dash">
            <Reports path="reports">
              <AnnualReport path="annual" />
            </Reports>
          </Dash>
        </Router>
      )
    });
  });

  it("renders at a path with nested paths", () => {
    snapshot({
      pathname: "/dash",
      element: (
        <Router>
          <Home path="/" />
          <Dash path="/dash">
            <Reports path="reports">
              <AnnualReport path="annual" />
            </Reports>
          </Dash>
        </Router>
      )
    });
  });

  it("renders a child 'index' nested path", () => {
    snapshot({
      pathname: "/dash",
      element: (
        <Router>
          <Home path="/" />
          <Dash path="/dash">
            <Reports path="/" />
          </Dash>
        </Router>
      )
    });
  });

  it("yo dawg", () => {
    snapshot({
      pathname: "/",
      element: (
        <Router>
          <Dash path="/">
            <Dash path="/">
              <Dash path="/" />
              <Reports path=":reportId" />
            </Dash>
          </Dash>
        </Router>
      )
    });
  });

  it("yo dawg again", () => {
    snapshot({
      pathname: "/",
      element: (
        <Router>
          <Dash path="/">
            <Dash path="/">
              <Dash path="/" />
              <Reports path="reports/:reportId" />
            </Dash>
          </Dash>
        </Router>
      )
    });
  });

  it("matches multiple nested / down to a child with a path", () => {
    snapshot({
      pathname: "/yo",
      element: (
        <Router>
          <Dash path="/">
            <Dash path="/">
              <Dash path="/yo" />
            </Dash>
          </Dash>
        </Router>
      )
    });
  });
});

describe("disrespect", () => {
  it("has complete disrespect for leading and trailing slashes", () => {
    snapshot({
      pathname: "dash/reports/annual/",
      element: (
        <Router>
          <Home path="/" />
          <Dash path="dash">
            <Reports path="/reports/">
              <AnnualReport path="annual" />
            </Reports>
          </Dash>
        </Router>
      )
    });
  });
});

describe("links", () => {
  it("accepts an innerRef prop", done => {
    let ref;
    let div = document.createElement("div");
    ReactDOM.render(
      <Link to="/" innerRef={node => (ref = node)} />,
      div,
      () => {
        expect(ref).toBeInstanceOf(HTMLAnchorElement);
        ReactDOM.unmountComponentAtNode(div);
        done();
      }
    );
  });

  it("forwards refs", done => {
    let ref;
    let div = document.createElement("div");
    ReactDOM.render(<Link to="/" ref={node => (ref = node)} />, div, () => {
      expect(ref).toBeInstanceOf(HTMLAnchorElement);
      ReactDOM.unmountComponentAtNode(div);
      done();
    });
  });

  it("renders links with relative hrefs", () => {
    const Parent = ({ children }) => (
      <div>
        <h1>Parent</h1>
        <Link to="reports">/dash/reports</Link>
        {children}
      </div>
    );

    const Child = () => (
      <div>
        <h2>Child</h2>
        <Link to="../">/dash</Link>
      </div>
    );

    snapshot({
      pathname: "/dash/reports",
      element: (
        <Router>
          <Parent path="dash">
            <Child path="reports" />
            <Child path="charts" />
          </Parent>
        </Router>
      )
    });
  });

  it("uses the right href in multiple root paths", () => {
    const Parent = ({ uri, children }) => (
      <div>
        <div>Parent URI: {uri}</div>
        {children}
      </div>
    );

    const Child = ({ uri }) => (
      <div>
        <div>Child URI: {uri}</div>
        <Link to="three">/one/two/three</Link>
        <Link to="..">/one</Link>
        <Link to="../..">/</Link>
      </div>
    );

    snapshot({
      pathname: "/one/two",
      element: (
        <Router>
          <Parent path="/">
            <Parent path="/">
              <Parent path="one">
                <Child path="two" />
              </Parent>
            </Parent>
          </Parent>
        </Router>
      )
    });
  });
});

describe("transitions", () => {
  it("transitions pages", async () => {
    const {
      snapshot,
      history: { navigate }
    } = runWithNavigation(
      <Router>
        <Home path="/" />
        <Reports path="reports" />
      </Router>
    );
    snapshot();
    await navigate("/reports");
    snapshot();
  });

  it("keeps the stack right on interrupted transitions", async () => {
    const {
      snapshot,
      history,
      history: { navigate }
    } = runWithNavigation(
      <Router>
        <Home path="/" />
        <Reports path="reports" />
        <AnnualReport path="annual-report" />
      </Router>
    );
    navigate("/reports");
    await navigate("/annual-report");
    snapshot();
    expect(history.index === 1);
  });
});

describe("relative navigate prop", () => {
  it("navigates relative", async () => {
    let relativeNavigate;

    const User = ({ children, navigate, userId }) => {
      relativeNavigate = navigate;
      return (
        <div>
          User:{userId}
          {children}
        </div>
      );
    };

    const Settings = () => <div>Settings</div>;

    const { snapshot } = runWithNavigation(
      <Router>
        <User path="user/:userId">
          <Settings path="settings" />
        </User>
      </Router>,
      "/user/123"
    );
    snapshot();
    await relativeNavigate("settings");
    snapshot();
  });

  it("navigates relative", () => {});
});

describe("nested routers", () => {
  it("allows arbitrary Router nesting through context", () => {
    const PageWithNestedApp = () => (
      <div>
        Home
        <ChatApp />
      </div>
    );

    const ChatApp = () => (
      <Router>
        <ChatHome path="/home" />
      </Router>
    );

    const ChatHome = () => <div>Chat Home</div>;

    snapshot({
      pathname: `/chat/home`,
      element: (
        <Router>
          <PageWithNestedApp path="/chat/*" />
        </Router>
      )
    });
  });
});

describe("Match", () => {
  it("matches a path", () => {
    snapshot({
      pathname: `/groups/123`,
      element: (
        <Match path="/groups/:groupId">
          {props => <PropsPrinter {...props} />}
        </Match>
      )
    });
  });
});

// React 16.4 is buggy https://github.com/facebook/react/issues/12968
describe.skip("ServerLocation", () => {
  let App = () => (
    <Router>
      <Home path="/" />
      <Group path="/groups/:groupId" />
      <Redirect from="/g/:groupId" to="/groups/:groupId" />
    </Router>
  );

  it("works", () => {
    expect(
      renderToString(
        <ServerLocation url="/">
          <App />
        </ServerLocation>
      )
    ).toMatchSnapshot();

    expect(
      renderToString(
        <ServerLocation url="/groups/123">
          <App />
        </ServerLocation>
      )
    ).toMatchSnapshot();
  });

  test("redirects", () => {
    let redirectedPath = "/g/123";
    let markup;
    try {
      markup = renderToString(
        <ServerLocation url={redirectedPath}>
          <App />
        </ServerLocation>
      );
    } catch (error) {
      expect(markup).not.toBeDefined();
      expect(isRedirect(error)).toBe(true);
      expect(error.uri).toBe("/groups/123");
    }
  });
});
