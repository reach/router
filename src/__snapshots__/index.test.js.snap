// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match matches a path 1`] = `
<pre>
  {
  "location": {
    "pathname": "/groups/123",
    "search": "",
    "state": null,
    "key": "initial"
  },
  "match": {
    "groupId": "123",
    "uri": "/groups/123",
    "path": "/groups/:groupId"
  }
}
</pre>
`;

exports[`Router children allows for fragments 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Annual Report
  </div>
</div>
`;

exports[`Router children ignores falsey chidlren 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Home
  </div>
</div>
`;

exports[`ServerLocation works 1`] = `"<div style=\\"outline:none\\" tabindex=\\"-1\\" role=\\"group\\"><div>Home</div></div>"`;

exports[`ServerLocation works 2`] = `"<div style=\\"outline:none\\" tabindex=\\"-1\\" role=\\"group\\" data-reactroot=\\"\\"><div>Group: <!-- -->123</div></div>"`;

exports[`disrespect has complete disrespect for leading and trailing slashes 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Reports 
        <div
          role="group"
          style={
            Object {
              "outline": "none",
            }
          }
          tabIndex="-1"
        >
          <div>
            Annual Report
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`links renders links with relative hrefs 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    <h1>
      Parent
    </h1>
    <a
      aria-current="page"
      href="/dash/reports"
      onClick={[Function]}
    >
      /dash/reports
    </a>
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        <h2>
          Child
        </h2>
        <a
          aria-current={undefined}
          href="/dash"
          onClick={[Function]}
        >
          /dash
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`links uses the right href in multiple root paths 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    <div>
      Parent URI: 
      /
    </div>
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        <div>
          Parent URI: 
          /
        </div>
        <div
          role="group"
          style={
            Object {
              "outline": "none",
            }
          }
          tabIndex="-1"
        >
          <div>
            <div>
              Parent URI: 
              /one
            </div>
            <div
              role="group"
              style={
                Object {
                  "outline": "none",
                }
              }
              tabIndex="-1"
            >
              <div>
                <div>
                  Child URI: 
                  /one/two
                </div>
                <a
                  aria-current={undefined}
                  href="/one/two/three"
                  onClick={[Function]}
                >
                  /one/two/three
                </a>
                <a
                  aria-current={undefined}
                  href="/one"
                  onClick={[Function]}
                >
                  /one
                </a>
                <a
                  aria-current={undefined}
                  href="/"
                  onClick={[Function]}
                >
                  /
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`location correctly parses pathname, search and hash fields 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    <div>
      location.pathname: [
      /print-location
      ]
    </div>
    <div>
      location.search: [
      ?it=works&with=queries
      ]
    </div>
  </div>
</div>
`;

exports[`nested rendering matches multiple nested / down to a child with a path 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Dash 
        <div
          role="group"
          style={
            Object {
              "outline": "none",
            }
          }
          tabIndex="-1"
        >
          <div>
            Dash 
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`nested rendering renders a child 'index' nested path 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Reports 
      </div>
    </div>
  </div>
</div>
`;

exports[`nested rendering renders a nested path 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Reports 
      </div>
    </div>
  </div>
</div>
`;

exports[`nested rendering renders a really nested path 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Reports 
        <div
          role="group"
          style={
            Object {
              "outline": "none",
            }
          }
          tabIndex="-1"
        >
          <div>
            Annual Report
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`nested rendering renders at a path with nested paths 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
  </div>
</div>
`;

exports[`nested rendering yo dawg 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Dash 
        <div
          role="group"
          style={
            Object {
              "outline": "none",
            }
          }
          tabIndex="-1"
        >
          <div>
            Dash 
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`nested rendering yo dawg again 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Dash 
        <div
          role="group"
          style={
            Object {
              "outline": "none",
            }
          }
          tabIndex="-1"
        >
          <div>
            Dash 
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`nested routers allows arbitrary Router nesting through context 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Home
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Chat Home
      </div>
    </div>
  </div>
</div>
`;

exports[`passed props parses dynamic segments and passes to components 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Group: 
    123
  </div>
</div>
`;

exports[`passed props parses multiple params when nested 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    123
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        123
         - 
        456
      </div>
    </div>
  </div>
</div>
`;

exports[`passed props passes the matched URI to the component 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <pre>
    {
  "path": "/groups/:groupId/users/:userId",
  "groupId": "123",
  "userId": "456",
  "uri": "/groups/123/users/456",
  "location": {
    "pathname": "/groups/123/users/456",
    "search": "",
    "state": null,
    "key": "initial"
  }
}
  </pre>
</div>
`;

exports[`passed props router location prop to nested path 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Reports 
      </div>
    </div>
  </div>
</div>
`;

exports[`passed props shadows params in nested paths 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Group: 
    burger
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Group: 
        milkshake
      </div>
    </div>
  </div>
</div>
`;

exports[`relative navigate prop navigates relative 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    User:
    123
  </div>
</div>
`;

exports[`relative navigate prop navigates relative 2`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    User:
    123
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div>
        Settings
      </div>
    </div>
  </div>
</div>
`;

exports[`route ranking / 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Root
  </div>
</div>
`;

exports[`route ranking /groups 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Groups
  </div>
</div>
`;

exports[`route ranking /groups/123 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Group Id: 
    123
  </div>
</div>
`;

exports[`route ranking /groups/123/users 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Users
  </div>
</div>
`;

exports[`route ranking /groups/123/users/456 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    User id: 
    456
    , Group Id: 
    123
  </div>
</div>
`;

exports[`route ranking /groups/123/users/a/bunch/of/junk 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Users Splat: 
  </div>
</div>
`;

exports[`route ranking /groups/123/users/me 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Me!
  </div>
</div>
`;

exports[`route ranking /groups/mine 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    MyGroup
  </div>
</div>
`;

exports[`route ranking /groups/mine/users 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    MyGroupUsers
  </div>
</div>
`;

exports[`route ranking /groups/mine/users/me 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Mine and Me!
  </div>
</div>
`;

exports[`route ranking /one/two/three/four/five 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Fiver 
    one
     
    two
     
    three
     
    four
     
    five
  </div>
</div>
`;

exports[`smoke tests renders at a path 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Dash 
  </div>
</div>
`;

exports[`smoke tests renders the root component at "/" 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Home
  </div>
</div>
`;

exports[`transitions keeps the stack right on interrupted transitions 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Annual Report
  </div>
</div>
`;

exports[`transitions transitions pages 1`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Home
  </div>
</div>
`;

exports[`transitions transitions pages 2`] = `
<div
  role="group"
  style={
    Object {
      "outline": "none",
    }
  }
  tabIndex="-1"
>
  <div>
    Reports 
  </div>
</div>
`;
